Make calico MTU configurable in bootstrap config

TeddyAndrieux opened the pull request:
**Component**:

'salt', 'kubernetes'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

#1095 

**Summary**:

- Allow multiple CIDRs for workload plane and control plane networks
- Make calico MTU configurable through bootstrap config
- Check MTU defined is smaller than the one on the local workload network interface

New bootstrap config network layout (still backward compatible with old style)

```
apiVersion: metalk8s.scality.com/v1alpha3
networks:
  workloadPlane:
    cidr:
      - 10.200.10.0/24
      - 10.200.20.0/24
    mtu: 4242
  controlPlane:
    cidr: 10.200.30.0/24
```

---
TODO:

- [x] Add unit test for `get_mtu_from_ip`
- [x] Add unit test for `get_ip_from_cidrs`
- [x] Update documentation for BootstrapConfig changes
- [x] Update tests (and vagrant) to use the new BootstrapConfig style
- [x] Add documentation about MTU configuration in bootstrap config
- [x] Add changelog entry 
 
---

Fixes: #1095

TeddyAndrieux committed:
salt: Add `get_mtu_from_ip` function in `metalk8s_network

Add a function in `metalk8s_network` salt module to retrieve MTU
configured on an interface from an IP (will be needed to check MTU on
workload plane network before configuring calico #1095)

TeddyAndrieux committed:
salt: Make Pod network `mtu` configurable

Add a `mtu` key in bootstrap configuration to configure the value for
MTU on Pod network, this value is used by calico to create the
interface for Pod network.
NOTE: We set the calico MTU to the workload MTU - 20 so that if IPinIP
is enabled the MTU is still valid

Fixes: #1095

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

slaperche-scality commented on salt/_modules/metalk8s_network.py:
```diff
@@ -79,3 +79,26 @@ def get_oidc_service_ip():
     return _pick_nth_service_ip(OIDC_ADDRESS_NUMBER)
+
+
+def get_mtu_from_ip(ip):
```
IP or CIDR?

Reading the [doc](https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.network.html#salt.modules.network.ifacestartswith), I would say CIDR.

TeddyAndrieux replied:
Not really the doc is wrong here it's not a cidr but an IP (maybe not full)

(And btw it's a bit bugged if you not specify the full IP because he consider 0 as wildcard)

```
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether fa:16:3e:2b:fd:d5 brd ff:ff:ff:ff:ff:ff
    inet 10.200.2.69/16 brd 10.200.255.255 scope global dynamic eth0
       valid_lft 71176sec preferred_lft 71176sec
    inet6 fe80::f816:3eff:fe2b:fdd5/64 scope link 
       valid_lft forever preferred_lft forever
3: tunl0@NONE: <NOARP,UP,LOWER_UP> mtu 1440 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
    inet 10.233.132.64/32 brd 10.233.132.64 scope global tunl0
       valid_lft forever preferred_lft forever

[root@bootstrap ~]# salt-call network.ifacestartswith 10.200.0.0/16
local:
[root@bootstrap ~]# salt-call network.ifacestartswith 10.200
local:
    - tunl0
    - eth0
[root@bootstrap ~]# salt-call network.ifacestartswith 10.200.
local:
    - eth0
```

gdemonet commented on salt/metalk8s/kubernetes/cni/calico/configured.sls:
```diff
@@ -3,6 +3,13 @@
 
 {%- set kubernetes_service_ip = salt.metalk8s_network.get_kubernetes_service_ip() %}
 
+{#- Check that service MTU configured is inferior to the local workload interface one #}
+{%- set workload_local_mtu = salt.metalk8s_network.get_mtu_from_ip(grains.metalk8s.workload_plane_ip) %}
+{%- if pillar['networks']['mtu'] > workload_local_mtu %}
+  {{ raise('Trying to configure CNI with ' ~ pillar['metalk8s']['networks']['mtu']
+           ~ 'MTU but local workload interface MTU is inferior ' ~ workload_local_mtu) }}
+{%- endif %}
```
Neat! Do you think we can add a similar check earlier in the bootstrap/`deploy_node` orchestrate?

Maybe consider adding these checks in `metalk8s.internal.preflight`?

TeddyAndrieux replied:
Yes I could add check earlier (maybe not in the preflight because to me preflight should be removed at some point, since it does not bring any value)

I will try to find a better place earlier I agree it fail a bit late here (at least for bootstrap because for new node it will fail when rendering the "bring new node to highstate" so not really late I think)

gdemonet replied:
Could be done later, and as you said, since we may want to do some refactoring around the pre-checks, would make sense to add as a task in #1999

NicolasT commented on salt/_pillar/metalk8s.py:
```diff
@@ -70,6 +71,7 @@ def _load_networks(config_data):
         'workload_plane': networks_data['workloadPlane'],
         'pod': networks_data.get('pods', DEFAULT_POD_NETWORK),
         'service': networks_data.get('services', DEFAULT_SERVICE_NETWORK),
+        'mtu': networks_data.get('mtu', DEFAULT_MTU)
```
I don't think this is at the right 'location' in the config. It's not something *next to* the pod, service and other network CIDRs. It is *the MTU of the workload plane network*, so it should go in there. Indeed, that means bumping the version of the config format.

TeddyAndrieux replied:
Right, as discussed, I will update it

NicolasT commented on salt/metalk8s/kubernetes/cni/calico/configured.sls:
```diff
@@ -3,6 +3,13 @@
+{%- set workload_local_mtu = salt.metalk8s_network.get_mtu_from_ip(grains.metalk8s.workload_plane_ip) %}
+{%- if pillar['networks']['mtu'] > workload_local_mtu %}
+  {{ raise('Trying to configure CNI with ' ~ pillar['metalk8s']['networks']['mtu']
+           ~ 'MTU but local workload interface MTU is inferior ' ~ workload_local_mtu) }}
```
"smaller", not "inferior".

TeddyAndrieux committed:
salt: Ability to provide several CIDRs for control and workload networks

In metalk8s we may have node not in the same L2 networks so that some
routing was needed to communicate between node and it's perfectly find
just we need to allow multiple CIDRs for workload plane and control
plane networks.
Now in bootstrap config you can provide a list of CIDRs for this 2
networks.
Bump apiVersion of bootstrap config as the layout of the file changed

TeddyAndrieux committed:
salt: Add module function to get the first IP available

Add a module function to retrieve the first IP available from several
CIDRs with an optionnal argument for `current_ip` so that we return this
current_ip if this IP still available in one of the provided CIDRs

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

TeddyAndrieux committed:
tests: Do not use networks pillar in salt-api tests

Before this commit we use networks pillar to compute the salt-api
address in salt-api tests, but it's wrong we should directly use the
MetalK8s grains containing the control plane IP we want.
Use this grains instead of computing the address from CIDR in the pillar

TeddyAndrieux committed:
vagrant: Use `v1alpha3` version of BootstrapConfiguration file

Use the new version `v1alpha3` for BootstrapConfiguration file
generated with vagrant

TeddyAndrieux committed:
eve: Use `v1alpha3` version of BootstrapConfiguration file

Use the new version `v1alpha3` for BootstrapConfiguration file
generated by openstack worker for multiple-nodes

TeddyAndrieux committed:
eve: Use `v1alpha3` version of BootstrapConfiguration file

Use the new version `v1alpha3` for BootstrapConfiguration file
generated by openstack worker for single-node rhel

gdemonet commented on salt/_modules/metalk8s_network.py:
```diff
@@ -79,3 +79,26 @@ def get_oidc_service_ip():
+
+    if len(ifaces) > 1:
+        log.warning(
+            'Several interfaces has "%s" as IP, %s will be used: %s',
```
```suggestion
            'Several interfaces match the IP "%s", %s will be used: %s',
```

gdemonet commented on salt/_pillar/metalk8s.py:
```diff
@@ -15,6 +15,7 @@
 
 DEFAULT_POD_NETWORK = '10.233.0.0/16'
 DEFAULT_SERVICE_NETWORK = '10.96.0.0/12'
+DEFAULT_MTU = 1460
```
That's the default Calico uses based on GCP. Should we keep it (to smooth upgrades), or find one that better suits our use cases? (for on-prem, I don't what the lowest reasonable MTU we'd find, but maybe we could start by considering the ones on Scality Cloud - 1500 apparently)

TeddyAndrieux replied:
Don't really know, but to me it's not really part of this PR, but indeed it's a good question

gdemonet replied:
Let's keep it for now, shouldn't hurt unless we start deploying in envs where this default value is too low.

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -90,15 +89,6 @@ def _check_node_status():
 # }}}
 # Helpers {{{
 
-def get_node_ip(node_name, ssh_config, bootstrap_config):
```
As discussed privately, this helper is here to get _another_ node's control plane IP, not the bootstrap one. So we can't just replace it with the `control_plane_ip` fixture, unfortunately :confused: 

Also, since we need it before installing Salt on this node, we cannot use the same approach (reading `grains`). Maybe we could still use Salt-SSH and the new `metalk8s_network.get_ip_from_cidrs` method, to avoid duplication of logic, but keeping the old one should also work, so that's your call :)

TeddyAndrieux replied:
A bit overkill to use salt-ssh to retrieve IP we will use for node (so for salt-ssh) to me.
Anyway I just consider that we have only 1 CIDR for control-plane network in our tests (which is acceptable I think, at least for the moment)

gdemonet reviewed (commented):
Some minor comments, but overall behaviour and user-facing features LGTM!

TeddyAndrieux committed:
tests: Add utils to get pillar key

Add a utils in tests to get a pillar key with salt

TeddyAndrieux committed:
tests: Use pillar instead of bootstrap config to retrive CP IP

In expansions test we need to have the CIDR of the control-plane network
to create the new nodes, instead of retrieving the CIDR from bootstrap
config directly, retrieve it from the pillar.

NOTE: In expansions test we expect to have only one CIDR for control
plane network

TeddyAndrieux committed:
salt/tests: Add unit tests for `get_ip_from_cidrs` salt function

Add unit tests for `get_ip_from_cidrs` function from `metalk8s_network`
salt custom module

TeddyAndrieux committed:
salt/tests: Add unit tests for `get_mtu_from_ip` salt function

Add unit tests for `get_mtu_from_ip` function from `metalk8s_network`
salt custom module

TeddyAndrieux committed:
docs: Update documentation about BootstrapConfig networks

BootstrapConfig layout just changed, this commit update the
documentation according to this Bootstrap config changes

TeddyAndrieux committed:
docs: Add documentation about workloadPlane MTU configuration

Add a small paragraph about MTU value that may be provided for
workloadPlane network in BootstrapConfiguration

Refs: #1095

gdemonet reviewed (dismissed):
LGTM!

TeddyAndrieux committed:
changelog: Add entry about multiple CIDRs support configurable MTU

Add a changelog entry about the ability to use multiple CIDRs for
workload and control plane networks and to configure the MTU used by
Calico

Refs: #1095

gdemonet commented on CHANGELOG.md:
```diff
@@ -10,6 +10,9 @@
   PR [#2369](https://github.com/scality/metalk8s/pull/2369))
+- [#1095](https://github.com/scality/metalk8s/issues/1095) - Ability to use
+  multiple CIDRs for control plane and workload plane networks and to configure
+  the MTU used by Calico (PR [#2677](https://github.com/scality/metalk8s/pull/2677))
```
Maybe specify "the workload plane MTU" as you did in docs?

gdemonet reviewed (dismissed):
Perfect! Thanks, I guess we just need to create some tracking tickets for open questions:
- how to get the MTU checks "standalone" so we can run them earlier
- what default MTU should we use
- what's the impact of having multiple CIDRs (manual + automated tests to consider)

No need to provide much details, just get a title, link to this PR, and the `state:question` label and we'll get back to them later.

NicolasT commented:
Something to keep in mind: when multiple CIDRs *are* used, you *almost* certainly require tunneling (IPIP/VXLAN) to be enabled! (Unless Calico is configured to peer with routers on the data path etc. which is rather unlikely in our deployments).

TeddyAndrieux commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/60877dc6e811185b9951836a860d720671f9d377) did not succeed in branch improvement/GH-1095-make-calico-mtu-configurable.




*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

TeddyAndrieux commented:
/wait

gdemonet commented on docs/installation/advanced-guide.rst:
```diff
@@ -0,0 +1,36 @@
+           - 10.100.2.0/28
+           - 10.200.2.0/28
+
+This kind of deployment need good knowledge about networking, as all workload
+node need to be able to communicate with all workload node even those that does
+not sit in the same CIDRs.
```
```suggestion
This kind of deployment needs good knowledge about networking, as each workload
node needs to be able to communicate with all others, even those in a different
workload CIDR.
```

gdemonet commented on docs/installation/bootstrap.rst:
```diff
@@ -60,11 +63,26 @@ notation for it's various subfields.
       host level for each member of the cluster.
 
+
+      .. warning::
```
Maybe let this just be an `info` admonition?

gdemonet commented on docs/installation/bootstrap.rst:
```diff
@@ -60,11 +63,26 @@ notation for it's various subfields.
+
+      .. warning::
+
+        Several CIDR can be provided if all nodes does not sit in the same
+        network but it's an
+        :ref:`advanced configuration<multiple CIDR network>`.
```
```suggestion
        Several CIDRs can be provided if all nodes do not sit in the same
        network. This is an ref:`advanced procedure<multiple CIDR network>`
        which we do not recommend for non-experts.
```

gdemonet reviewed (changes requested):
Minor wording comments, and we should be good to go!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

TeddyAndrieux commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue GH-1095.

Goodbye teddyandrieux.
